vmcnet/updates/params.py

```diff
@@ -190,6 +211,13 @@ def update_param_fn(params, data, optimizer_state, key):
 
         if record_param_l1_norm:
             metrics.update({"param_l1_norm": stats_to_save[4]})
+        if spin_square_expectation_fn is not None:
+            spin_squared_expectation = traced_compute_spin_square(params, positions)
+            if optimizer.multi_device:
+                spin_squared_expectation = utils.distribute.get_first(
```

Comment:
The get_first is definitely necessary to have a single float output, but we could compose it with a reduction over all devices to get the overall expectation to be more accurate. My one reservation is that avoiding the communication between devices might be more important than the sqrt(ndevices) accuracy that you gain by doing the pmean_if_pmap stuff, since the spin is less critical to the optimization than stuff like the energy/variance.

Reply:
Mm yeah not sure, feels somewhat nice to use all the data we have but I can also see an argument for just using one device to save time. I'm fine with either way.